### PR TITLE
feat(stdlib): SMT with foreign merkle_membership

### LINKED ISSUE #1080: Add convenience structure for sparse merkle trees
### Problem

We currently have a method for checking membership, however it is not intuitive as to how one would check for non-membership using it. 

**More details:**

A merkle tree is mainly used for set membership proofs. So with it i will be able to tell you if an element x was a part of a set in an efficient way within circuits.

For example, if your set is `S = {1,2,3,4}`  and `x=5` then `check_membership(S, x)` would return false.

If `x=2` then `check_membership(S, x)`  would return true.

**Regular merkle tree**

A regular merkle tree is allowed to have a set that looks like this `SA = {200,8,1,1,2,3,4,5,6,7,8,8,8,8,}`

Notice that there are duplicates. This is called an append-only merkle tree, because you can just keep on adding any element you want into the tree. More importantly, notice that an element does not have a "rightful place" in the set.

If I do `check_membership(SA, 1)` this will return true though there are multiple values of 1 in SA so I could reference any one.

**Non-membership**

Since each element does not have a "rightful place" and duplicates are allowed, I cannot prove that an element is not in the set, without looking at every element in the set. This is very in-efficient, and grows with the number of elements in the set.

**Sparse merkle tree**

The set for a sparse merkle tree on the other hand will look like this: `SS = {1, empty_2, 3, empty_4, empty_5, empty_6, empty_7, 8}`
I can also do `check_membership(SS, 1)` and this will return true, but there is only one place that the number 1  can be in. This is just a property of a sparse merkle tree data structure, Hence there are no other 1's  in this set since that is how we have chosen to insert data in the underlying database.

I can additionally prove that an element is not in the set quite easily. Lets say I want to show that the number 2 is not in the set. I just need to do  `check_membership(SS, empty_2)` . If that spot is empty, then I know that 2 is not in the set because the underlying data structure will not insert 2  in any other place.

**Noir**

Notice that in Noir, we don't have a separate function for each of these trees. We still only use check_membership . The catch here is that we need to know what the default value is when something is not in the set. Its usually 0 but in the proposed solution, the user can set it to whatever value they see fit.

### Proposed solution

Using pseudo-code I propose we make a struct:

```rust
struct SparseMerkleTree {
 default_value : Field,
}

impl SparseMerkleTree {
 fn update(self) {}
 fn delete(self) {}
 fn insert(self) {}
}
```

### Alternatives considered

_No response_

### Additional context

I think we can have a helper standalone update method in the stdlib that gives a merkle update proof.

This would shows that some value `v1` was in the tree with root `R1` , then when I update `v1` to `v2` , the root becomes `R2`

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by joss-aztec**
# Related issue(s)

Resolves #1080

# Description

## Summary of changes

Simple helper for implementing sparse merkle tree operations in terms of the black box foreign function `check_membership`.

## Test additions / changes

# Checklist

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [ ] I have reviewed the changes on GitHub, line by line.
- [ ] I have ensured all changes are covered in the description.

## Documentation needs
- [ ] This PR requires documentation updates when merged.

# Additional context

### CODE REVIEW

File: noir_stdlib/src/merkle/sparse_merkle_tree.nr [hunk 1]
```diff
@@ -0,0 +1,84 @@
+use crate::merkle;
+
```
Comment (kevaundray) [inline noir_stdlib/src/merkle/sparse_merkle_tree.nr:2]: Can we add some comments to the functions here and is there a way to test functionality?

Comment (joss-aztec) [reply noir_stdlib/src/merkle/sparse_merkle_tree.nr:2]: Sure. Relatedly, I've not noticed any explicit comments/docs on the hash `check_membership` expects the tree to have used. This should probably be added too. I'm guessing it's pedersen?

Comment (kevaundray) [reply noir_stdlib/src/merkle/sparse_merkle_tree.nr:2]: Yeah that would be great!

Comment (joss-aztec) [reply noir_stdlib/src/merkle/sparse_merkle_tree.nr]: Added

Comment (joss-aztec) [reply noir_stdlib/src/merkle/sparse_merkle_tree.nr]: Spoke too soon. The test proves, but doesn't verify D:
Guessing this test has exposed a bug that deserves a separate issue?
```
Failed to verify proof
thread 'tests::noir_integration' panicked at 'verification fail for "sparse_merkle_tree"', crates/nargo_cli/tests/prove_and_verify.rs:61:25
```

Comment (joss-aztec) [reply noir_stdlib/src/merkle/sparse_merkle_tree.nr]: Just noticed when proving from nargo it logs:
```
linear_eval is not 0.
```

File: noir_stdlib/src/merkle/smt.nr [hunk 2]
```diff
@@ -0,0 +1,73 @@
+use crate::merkle;
+
+struct SparseMerkleTree {
+    default_value: Field,
+}
+
+impl SparseMerkleTree {
+    /// Constrains that the new leaf is located in the new tree at an index
```
Comment (kevaundray) [inline noir_stdlib/src/merkle/smt.nr:8]: we don't have doc comments in noir yet, only regular comments

File: noir_stdlib/src/merkle/sparse_merkle_tree.nr (hunk 1 above)
Comment (kevaundray) [reply noir_stdlib/src/merkle/sparse_merkle_tree.nr]: I think we have an issue with a similar bug, can you link this PR to it as another corpus?

Comment (kevaundray) [reply noir_stdlib/src/merkle/sparse_merkle_tree.nr]: Would also be good if you could add a stripped down example into the program (We can revisit this and the other issue once we switch to ultra plonk)

File: noir_stdlib/src/merkle/smt.nr (hunk 2 above)
Comment (joss-aztec) [reply noir_stdlib/src/merkle/smt.nr]: Now changed to regular comments

File: crates/nargo_cli/tests/test_data/sparse_merkle_tree/Nargo.toml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+
+        [package]
+        authors = [""]
+        compiler_version = "0.1"
+    
+        [dependencies]
+
```
Comment (TomAFrench) [inline crates/nargo_cli/tests/test_data/sparse_merkle_tree/Nargo.toml:7]:
```suggestion
[package]
authors = [""]
compiler_version = "0.1"

[dependencies]
```

**Comment by kevaundray** [comment]
Tracking issue : https://github.com/noir-lang/noir/issues/1252

**Review by TomAFrench — CHANGES_REQUESTED** [review]
We probably shouldn't include this in the stdlib for reasons mentioned in https://github.com/noir-lang/noir/issues/1258

**Comment by TomAFrench** [comment]
Closing this PR for thin stdlib reasons and it's out of date..

**Pull request closed without merging**